Modules/SaltStackConfig/SaltStackConfig.psm1

```diff
@@ -0,0 +1,304 @@
+<#
+Copyright 2021 VMware, Inc.
+SPDX-License-Identifier: BSD-2-Clause
+#>
+Function Connect-SscServer {
+<#
+  .NOTES
+  ===========================================================================
+   Created by:	Brian Wuchner
+   Date:		November 27, 2021
+   Blog:		www.enterpriseadmins.org
+   Twitter:		@bwuch
+  ===========================================================================
+  .SYNOPSIS
+    This function will allow you to connect to a vRealize Automation SaltStack Config API.
+    A global variable will be set with the Servername & Cookie/Header value for use by other functions.
+  .DESCRIPTION
+    Use this function to create the cookie/header to connect to SaltStack Config
+  .EXAMPLE
+    PS C:\> Connect-SscServer -Server 'salt.example.com' -Username 'root' -Password 'VMware1!'
+    This will default to internal user authentication.
+  .EXAMPLE
+    PS C:\> Connect-SscServer -Server 'salt.example.com' -Username 'bwuchner' -Password 'MyPassword1!' -AuthSource 'LAB Directory'
+    This will use the 'Lab Directory' LDAP authentication source.
+#>
+  param(
+    [Parameter(Mandatory=$true)][string]$server,
+    [Parameter(Mandatory=$true)][string]$username,
+    [Parameter(Mandatory=$true)][ValidateNotNullOrEmpty()][string]$password,
+    [string]$AuthSource='internal'
+  )
+  
+  $loginBody = @{'username'=$username; 'password'=$password; 'config_name'=$AuthSource}
+  try {
+    $webRequest = Invoke-WebRequest -Uri "https://$server/account/login" -SessionVariable ws -Headers $header
+    $ws.headers.Add('X-Xsrftoken', $webRequest.headers.'x-xsrftoken')

```

Comment:
Agreed this looks odd, but in testing its the only way it seemed to work.  It is similar to what is observed in the UI (using web browser developer tools) and the official documentation with a curl example (included below).  There is first a curl get to return a cookie, then that cookie is parsed for the X-Xsrftoken and used as a header in a subsequent post.  In my example, I'm only sending username/password in the POST so that I can also include an Auth Source (config_name) if needed to support something like LDAP authentication. 

curl example for API:
```
curl -k -c $HOME/eAPICookie.txt -u root:PASSWORD 'https://localhost/account/login' >/
˓→dev/null
curl -k -u root:PASSWORD -b $HOME/eAPICookie.txt \
-H 'X-Xsrftoken: '$(grep -w '_xsrf' $HOME/eAPICookie.txt | cut -f7)'' \
-X POST https://localhost/rpc \
-d '{
"resource": "sec",
"method": "download_content",
"kwarg": {"auto_ingest": true}
}'
```

Reply:
What I'm seeing on the curl example is very different. They are only making one call to /account/login and they are passing the username and password using basic authentication. The next call after that is an /rpc call, so here only the first call makes the authentication. The PowerShell equivalent of this example would be

$secStringPassword = ConvertTo-SecureString $password -AsPlainText -Force
$credObject = New-Object System.Management.Automation.PSCredential ($userName, $secStringPassword)
Invoke-WebRequest -Uri "https://$server/account/login" -SkipCertificateCheck -Authentication Basic -Credential $credObject

Reply:
Looking back through my notes, I had something very similar to this initially.  It worked great for local user/pass authentication, but every POST to pass in a 'config_name' element when using LDAP as the authentication source resulted in a `The remote server returned an error: (403) Forbidden.` error message.  Watching what happens in browser developer tools, I see that the X-Xsrftoken is populated in the request header prior to doing the POST operation, so I put what you see here together to get that header value prior to doing a post and things started working.  I'm happy to try other ideas, but for my use case this LDAP account functionality is important & I need to keep it working.  If it makes things better, I could do the clean method you suggested when the AuthSource parameter is internal and only go down this odd path when AuthSource is specified.  Would that be any better?

Reply:
Isn't there any documentation or curl sample on how to use the API with LDAP account?

Reply:
I have found the documentation and it really doesn't mention how to login with LDAP account. So you're saying that if you do 
Invoke-WebRequest -Uri "https://$server/account/login" -method POST -body (ConvertTo-Json $loginBody) without populating X-Xsrftoken before that this call fails with 403 Forbidden? It is quite odd for a login method to require a token. But if that how the API is implemented I'm ok with the your code.

Reply:
That is correct.

The codeblock:
```
$server = 'salt.example.com'
$loginBody = @{'username'='readonly'; 'password'='VMware1!'; 'config_name'='LAB Directory'}
Invoke-WebRequest -Uri "https://$server/account/login" -method POST -body (ConvertTo-Json $loginBody)
```
Returns:
```
Invoke-WebRequest : The remote server returned an error: (403) Forbidden.
At line:1 char:1
+ Invoke-WebRequest -Uri "https://$server/account/login" -method POST - ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExc
   eption
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```

However, the two call method with one GET to return the X-Xsrftoken and a second POST to pass in config_name will connect.  The codeblock:
```
$server = 'salt.example.com'
$loginBody = @{'username'='readonly'; 'password'='VMware1!'; 'config_name'='LAB Directory'}
$webRequest = Invoke-WebRequest -Uri "https://$server/account/login" -SessionVariable ws
$ws.headers.Add('X-Xsrftoken', $webRequest.headers.'x-xsrftoken')
Invoke-WebRequest -Uri "https://$server/account/login" -WebSession $ws -method POST -body (ConvertTo-Json $loginBody)
```
Returns:
```
StatusCode        : 200
StatusDescription : OK
Content           : {"type": "account", "id": "663b1203-1a5e-40c0-81f2-09d28eb31b46", "authenticated": true,
                    "attributes": {"config_name": "LAB Directory", "config_driver": "ldap", "username": "readonly",
                    "roles": {"faa9...
RawContent        : HTTP/1.1 200 OK
                    X-Xsrftoken: 2|83243e72|1e94fe0f728521e62d3a32c05c4eefdc|1638464453
                    Raas-Version: 8.6.0
                    Raas-Instance-Id: 35066a3b-84d6-461d-84c8-151adb91a143
                    Content-Security-Policy: script-src '...
Forms             : {}
Headers           : {[X-Xsrftoken, 2|83243e72|1e94fe0f728521e62d3a32c05c4eefdc|1638464453], [Raas-Version, 8.6.0],
                    [Raas-Instance-Id, 35066a3b-84d6-461d-84c8-151adb91a143], [Content-Security-Policy, script-src
                    'self' 'unsafe-inline' 'unsafe-eval' blob:; frame-ancestors 'self'; object-src 'none']...}
Images            : {}
InputFields       : {}
Links             : {}
ParsedHtml        : mshtml.HTMLDocumentClass
RawContentLength  : 656
```

I've reached out to a few folks asking if there is a better example, but if you are ok with this for now I will fix it later if I find a better way.  

Resolving this conversation for now.